[DependencyInjection] deprecate the `@required` annotation

alexislefebvre opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #48909
| License       | MIT
| Doc PR        | -

TODO:

- [x] check the name of the package which should be reported in the deprecation message:
  - `symfony/framework-bundle` (that call `addGlobalIgnoredName('required')`)?
  - and/or
  - `symfony/dependency-injection` (that detect and is called when there is this annotation)?
- [x] fix or remove the test of the deprecation? ` $this->expectDeprecation` doesn't work as I expected

alexislefebvre committed:
[DependencyInjection] deprecate the `@required` annotation

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.php:
```diff
@@ -21,6 +21,7 @@
 
 return static function (ContainerConfigurator $container) {
     $container->services()
+        // remove this in Symfony 7.0, the “@required” annotation will be removed
```
please use ASCII double quotes

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -56,7 +56,10 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                     break;
                 }
                 if (false !== $doc = $r->getDocComment()) {
+                    // remove this in Symfony 7.0, the “@required” annotation will be removed
```
ASCII

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -56,7 +56,10 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                 if (false !== $doc = $r->getDocComment()) {
+                    // remove this in Symfony 7.0, the “@required” annotation will be removed
                     if (false !== stripos($doc, '@required') && preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@required(?:\s|\*/$)#i', $doc)) {
+                        trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation will be removed. Use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
will be removed. Use the => is deprecated, use the

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:
```diff
@@ -21,6 +21,8 @@
  * @author Sebastien Morel (Plopix) <morel.seb@gmail.com>
  * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @group legacy
```
should be removed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:
```diff
@@ -49,6 +51,8 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                 continue;
             }
 
+            trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation will be removed. Use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
Let's move this before the array_key_exists +

will be removed. Use => is deprecated, use

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ CHANGELOG
  * Deprecate `framework:exceptions` tag, unwrap it and replace `framework:exception` tags' `name` attribute by `class`
  * Deprecate the `notifier.logger_notification_listener` service, use the `notifier.notification_logger_listener` service instead
  * Allow setting private services with the test container
+ * Deprecate the `@required` annotation, use the attribute `Symfony\Contracts\Service\Attribute\Required` instead
```
```suggestion
 * Deprecate the `@required` annotation, use the `Symfony\Contracts\Service\Attribute\Required` attribute instead
```

OskarStark commented on src/Symfony/Component/DependencyInjection/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ CHANGELOG
  * Deprecate undefined and numeric keys with `service_locator` config
  * Fail if Target attribute does not exist during compilation
  * Enable deprecating parameters with `ContainerBuilder::deprecateParameter()`
+ * Deprecate the `@required` annotation, use the attribute `Symfony\Contracts\Service\Attribute\Required` instead
```
```suggestion
 * Deprecate the `@required` annotation, use the `Symfony\Contracts\Service\Attribute\Required` attribute instead
```

OskarStark commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -56,7 +56,10 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                 if (false !== $doc = $r->getDocComment()) {
+                    // remove this in Symfony 7.0, the "@required" annotation will be removed
                     if (false !== stripos($doc, '@required') && preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@required(?:\s|\*/$)#i', $doc)) {
+                        trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation is deprecated, use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
```suggestion
                        trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```

OskarStark commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:
```diff
@@ -45,6 +45,7 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
             ) {
                 continue;
             }
+            trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation is deprecated, use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
```suggestion
            trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```

OskarStark commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:
```diff
@@ -766,8 +769,13 @@ public function testWithNonExistingSetterAndAutowiring()
+     */
     public function testExplicitMethodInjection()
     {
+        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation will be removed. Use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
```suggestion
        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation will be removed. Use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```

OskarStark commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowireRequiredMethodsPassTest.php:
```diff
@@ -70,8 +73,13 @@ public function testSetterInjectionWithAttribute()
+     */
     public function testExplicitMethodInjection()
     {
+        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation will be removed. Use the attribute "Symfony\Contracts\Service\Attribute\Required" instead.');
```
```suggestion
        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation will be removed. Use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```

alexislefebvre commented:
Thanks for the reviews! Suggestions have been proceeded.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.php:
```diff
@@ -21,6 +21,7 @@
 
 return static function (ContainerConfigurator $container) {
     $container->services()
+        // remove this in Symfony 7.0, the "@required" annotation will be removed
```
`@deprecated since Symfony 6.3, to be removed in 7.0` to help with grepping when preparing 7.0

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -56,7 +56,10 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                     break;
                 }
                 if (false !== $doc = $r->getDocComment()) {
+                    // remove this in Symfony 7.0, the "@required" annotation will be removed
```
should be removed, the deprecation below is enough

alexislefebvre commented:
I tried to improve the deprecation by showing the source of the deprecation, and fixed a bug that reported a deprecation when using `#[Required]`.

Hopefully I fixed the tests, I duplicated some of them, in order to test `@required` and `#[Required]`, but I kept since the PR was growing quickly. Do we need to cover each of these scenarios?

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:
```diff
@@ -40,11 +40,16 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
             }
+            // Trigger a deprecation only when PHPDoc has been found.
+            if ($doc) {
+                trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead. It was found on "%s::%s".', $value::class, $reflectionProperty->getName());
```
`instead on property "%s::$%s".`

alexisLefebvre replied:
Thanks for the suggestion, I updated it without `instead`.

nicolas-grekas reviewed (commented):
The `@group legacy` annotation should be added only on test cases that should be removed on v7. I don't think that's the case right now.
Instead of adding it to many test cases, we should update them to use the attribute, and ensure we keep at least one test ensuring that the annotation still works.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:
```diff
@@ -40,11 +40,16 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
             if (!($type = $reflectionProperty->getType()) instanceof \ReflectionNamedType) {
                 continue;
             }
+            $doc = null;
```
$doc = false do reduce the range of possible types

alexislefebvre commented:
I didn't know how to decide if only one test with `@required` was enough, so I duplicated tests to use `@required` annotation or `#[Required]`attribute. This should prevent BC breaks on the 6.x branch.

But please let me know if some cases can be dropped.

nicolas-grekas reviewed (changes requested):
On second look, this misses at least the following patch. It's not enought because then you'll see 

```
  4x: Since symfony/dependency-injection 6.3: The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.
    2x in AutowirePassTest::testExplicitMethodInjectionAttribute from Symfony\Component\DependencyInjection\Tests\Compiler
    2x in AutowireRequiredMethodsPassTest::testExplicitMethodInjectionAttribute from Symfony\Component\DependencyInjection\Tests\Compiler
```

But this is legit: testExplicitMethodInjectionAttribute is currently using `SetterInjection::class`, which is itself deprecated. The test needs to be updated to not use any deprecated class.

```diff
diff --git a/src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php b/src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php
index 53ae32af85..bf1954a00e 100644
--- a/src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php
+++ b/src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php
@@ -69,6 +69,7 @@ class AutowireRequiredMethodsPass extends AbstractRecursivePass
                     if (false === stripos($doc, '@inheritdoc') || !preg_match('#(?:^/\*\*|\n\s*+\*)\s*+(?:\{@inheritdoc\}|@inheritdoc)(?:\s|\*/$)#i', $doc)) {
                         break;
                     }
+                    trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
                 }
                 try {
                     $r = $r->getPrototype();
diff --git a/src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php b/src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php
index 42ce6c6ee1..34ff97ce1e 100644
--- a/src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php
@@ -947,13 +947,9 @@ class AutowirePassTest extends TestCase
 
     /**
      * @dataProvider provideNotWireableCalls
-     *
-     * @group legacy
      */
     public function testNotWireableCalls($method, $expectedMsg)
     {
-        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
-
         $container = new ContainerBuilder();
 
         $foo = $container->register('foo', NotWireable::class)->setAutowired(true)
@@ -986,13 +982,8 @@ class AutowirePassTest extends TestCase
         ];
     }
 
-    /**
-     * @group legacy
-     */
     public function testSuggestRegisteredServicesWithSimilarCase()
     {
-        $this->expectDeprecation('Since symfony/dependency-injection 6.3: The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
-
         $container = new ContainerBuilder();
 
         $container->register(LesTilleuls::class, LesTilleuls::class);
diff --git a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php
index 3083be42c2..4a2c1f457f 100644
--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/includes/autowiring_classes.php
@@ -446,8 +446,7 @@ class NotWireable
     {
     }
 
-    // @deprecated since Symfony 6.3, to be removed in 7.0
-    /** @required */
+    #[Required]
     protected function setProtectedMethod(A $a)
     {
     }
```

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.php:
```diff
@@ -21,6 +21,7 @@
 
 return static function (ContainerConfigurator $container) {
     $container->services()
+        // @deprecated since Symfony 6.3, to be removed in 7.0
```
shouldn't this related only to the `addGlobalIgnoredName` call and not to the service definition ?

alexisLefebvre replied:
This comment was meant to apply only to the next 2 lines:

```php
        ->set('annotations.reader', AnnotationReader::class)
        ->call('addGlobalIgnoredName', ['required'])
```

But you're right, the service definition must be kept in the future. I'll change this, thanks.

stof commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -67,6 +69,7 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                     if (false === stripos($doc, '@inheritdoc') || !preg_match('#(?:^/\*\*|\n\s*+\*)\s*+(?:\{@inheritdoc\}|@inheritdoc)(?:\s|\*/$)#i', $doc)) {
                         break;
                     }
+                    trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```
this one looks wrong to me. It will trigger the deprecation for any method using `@inheritdoc`, which is totally wrong

alexisLefebvre replied:
I had the same question: https://github.com/symfony/symfony/pull/48990#issuecomment-1408394474

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.php:
```diff
@@ -22,6 +22,7 @@
 return static function (ContainerConfigurator $container) {
     $container->services()
         ->set('annotations.reader', AnnotationReader::class)
+            // @deprecated since Symfony 6.3, to be removed in 7.0
```
This should not even be removed in 7.0. The deprecation warning is not triggered when using both the attribute and the annotation. So when migrating to 7.0, we won't be in a case where you have a code base without `@required` (your codebase won't rely on it being effective but it can still have it to support multiple versions of Symfony). And so we cannot stop telling the Doctrine annotation reader to ignore it.

alexisLefebvre commented on src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:
```diff
@@ -67,6 +69,7 @@ protected function processValue(mixed $value, bool $isRoot = false): mixed
                     if (false === stripos($doc, '@inheritdoc') || !preg_match('#(?:^/\*\*|\n\s*+\*)\s*+(?:\{@inheritdoc\}|@inheritdoc)(?:\s|\*/$)#i', $doc)) {
                         break;
                     }
+                    trigger_deprecation('symfony/dependency-injection', '6.3', 'The "@required" annotation on methods is deprecated, use the "Symfony\Contracts\Service\Attribute\Required" attribute instead.');
```
```suggestion
```

Let's see if CI passes without the second `trigger_deprecation`

alexislefebvre commented:
Status: Needs Review

nicolas-grekas commented:
Thank you @alexislefebvre.
